addons/sale_timesheet/models/sale_order.py

```diff
@@ -72,6 +72,10 @@ def _compute_field_value(self, field):
                 # We want to display only one time the warning for each SOL
                 upsellable_lines.write({'has_displayed_warning_upsell': True})
 
+    def _compute_visible_hours_recorded_button(self):
+        for order in self:
+            order.visible_hours_recorded_button = order.order_line.filtered(lambda l: l.is_service and l.product_id.service_policy in ['ordered_prepaid', 'delivered_timesheet']) and not order.state in ['draft', 'sent']
```

Comment:
This could be optimized. I would at least check the `order.state` first, and there might be further improvements with product.product caching (especially considering that `l.is_service` is basically `l.product_id.type == 'service'`).

Reply:
The order.state part is straigth-forward, but i'm not sure to know what you mean/want with that cache thing. I mean, i know what  a cache memory is, but i dont know how to use it in odoo.